### PR TITLE
`azurerm_mssql_virtual_machine` - Add support for `storage_configuration_settings`

### LINKED ISSUE #6874: Support for SQL Virtual Machine StorageConfigurationSettings
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Currently you're not able to add StorageConfigurationSettings to the `azurerm_mssql_virtual_machine`. We're changing some values in that configuration in our ARM template currently so I'm not able to switch to the resource itself. The SDK supports it so it just needs to be implemented I guess.

Example excerpt in ARM:
```json
...
"StorageConfigurationSettings": {
  "DiskConfigurationType": "NEW",
  "StorageWorkloadType": "OLTP",
  "SQLDataSettings": {
    "LUNs": [0],
    "DefaultFilePath": "F:\\data"
  },
  "SQLLogSettings": {
    "LUNs": [0],
    "DefaultFilePath": "F:\\log"
  },
  "SQLTempDbSettings": {
    "DefaultFilePath": "D:\\tempDb"
  }
},
...
```

### New or Affected Resource(s)

* azurerm_mssql_virtual_machine

### Potential Terraform Configuration

Rough idea below:

```hcl
...
storage_configuration_settings {
  disk_configuration_type = "NEW"
  workload_type           = "OLTP"

  sql_data_luns              = [0]
  sql_data_default_file_path = "F:\\data"

  sql_logs_luns             = [0]
  sql_log_default_file_path = "F:\\log"

  sql_temp_db_luns             = [0]
  sql_tem_db_default_file_path = "D:\\tempDb"
}
...
```

Not sure if we really need a block here or if we can simply put all the items as top level properties. Needs to be figured out. We can potentially even get rid of the `sql_` prefixes for the LUNs and file paths.

### References

* https://github.com/Azure/azure-sdk-for-go/blob/master/services/preview/sqlvirtualmachine/mgmt/2017-03-01-preview/sqlvirtualmachine/models.go#L1649

### PR DESCRIPTION
**Pull request by mbfrahry**
Fixes #6874

### CODE REVIEW

File: azurerm/internal/services/mssql/mssql_virtual_machine_resource.go [hunk 1]
```diff
@@ -171,6 +172,37 @@ func resourceArmMsSqlVirtualMachine() *schema.Resource {
 				ValidateFunc: validate.MsSqlVMLoginUserName,
 			},
 
+			"storage_configuration_settings": {
```
Comment (katbyte) [inline azurerm/internal/services/mssql/mssql_virtual_machine_resource.go:175]: I think settings is redundant here
```suggestion
			"storage_configuration": {
```

File: azurerm/internal/services/mssql/mssql_virtual_machine_resource.go [hunk 2]
```diff
@@ -171,6 +172,37 @@ func resourceArmMsSqlVirtualMachine() *schema.Resource {
...
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"disk_configuration_type": {
```
Comment (katbyte) [inline azurerm/internal/services/mssql/mssql_virtual_machine_resource.go:181]: as is configuration?
```suggestion
						"disk_type": {
```

File: azurerm/internal/services/mssql/mssql_virtual_machine_resource.go [hunk 3]
```diff
@@ -171,6 +172,37 @@ func resourceArmMsSqlVirtualMachine() *schema.Resource {
...
+						"storage_workload_type": {
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: validation.StringInSlice([]string{
+								string(sqlvirtualmachine.GENERAL),
+								string(sqlvirtualmachine.OLTP),
+								string(sqlvirtualmachine.DW),
+							}, false),
+						},
+						"sql_data_settings":    helper.StorageSettingSchema(),
```
Comment (katbyte) [inline azurerm/internal/services/mssql/mssql_virtual_machine_resource.go:199]: as is sql?
```suggestion
						"data_settings":    helper.StorageSettingSchema(),
```

File: azurerm/internal/services/mssql/mssql_virtual_machine_resource.go [hunk 4]
```diff
@@ -171,6 +172,37 @@ func resourceArmMsSqlVirtualMachine() *schema.Resource {
...
+						"sql_log_settings":     helper.StorageSettingSchema(),
```
Comment (katbyte) [inline azurerm/internal/services/mssql/mssql_virtual_machine_resource.go:200]:
```suggestion
						"log_settings":     helper.StorageSettingSchema(),
```

File: azurerm/internal/services/mssql/mssql_virtual_machine_resource.go [hunk 5]
```diff
@@ -171,6 +172,37 @@ func resourceArmMsSqlVirtualMachine() *schema.Resource {
...
+						"sql_temp_db_settings": helper.StorageSettingSchema(),
```
Comment (katbyte) [inline azurerm/internal/services/mssql/mssql_virtual_machine_resource.go:201]:
```suggestion
						"temp_db_settings": helper.StorageSettingSchema(),
```

**Review by katbyte — CHANGES_REQUESTED** [review]
Aside from some comments about the property this is looking good

**Comment by a30000931** [comment]
looks great - happy to help test this out

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @mbfrahry - looks like the tests are failing thou:

```	[TestAccAzureRMMsSqlVirtualMachine_storageConfigurationSettings] [Test Output]
=== RUN   TestAccAzureRMMsSqlVirtualMachine_storageConfigurationSettings
=== PAUSE TestAccAzureRMMsSqlVirtualMachine_storageConfigurationSettings
=== CONT  TestAccAzureRMMsSqlVirtualMachine_storageConfigurationSettings
    TestAccAzureRMMsSqlVirtualMachine_updateAutoPatching: testing.go:684: Step 0 error: Check failed: Check 2/4 error: azurerm_mssql_virtual_machine.test: Attribute 'auto_patching.0.day_of_week' expected "Sunday", got ""
    TestAccAzureRMMsSqlVirtualMachine_complete: testing.go:684: Step 0 error: After applying this step, the plan was not empty:
        
        DIFF:
        
        UPDATE: azurerm_mssql_virtual_machine.test
          auto_patching.#:                               "0" => "0"
          id:                                            "/subscriptions/*******/resourceGroups/acctestRG-mssql-200929043203782538/providers/Microsoft.SqlVirtualMachine/sqlVirtualMachines/acctest-VM-200929043203782538" => "/subscriptions/*******/resourceGroups/acctestRG-mssql-200929043203782538/providers/Microsoft.SqlVirtualMachine/sqlVirtualMachines/acctest-VM-200929043203782538"
          key_vault_credential.#:                        "0" => "0"
          r_services_enabled:                            "true" => "true"
          sql_connectivity_port:                         "1433" => "1433"
          sql_connectivity_type:                         "PRIVATE" => "PRIVATE"
          sql_connectivity_update_password:              "Password1234!" => "Password1234!"
          sql_connectivity_update_username:              "sqllogin" => "sqllogin"
          sql_license_type:                              "PAYG" => "PAYG"
          storage_configuration.#:                       "1" => "0"
          storage_configuration.0.data_settings.#:       "0" => ""
          storage_configuration.0.disk_type:             "" => ""
          storage_configuration.0.log_settings.#:        "0" => ""
          storage_configuration.0.storage_workload_type: "" => ""
          storage_configuration.0.temp_db_settings.#:    "0" => ""
          virtual_machine_id:                            "/subscriptions/*******/resourceGroups/acctestRG-mssql-200929043203782538/providers/Microsoft.Compute/virtualMachines/acctest-VM-200929043203782538" => "/subscriptions/*******/resourceGroups/acctestRG-mssql-200929043203782538/providers/Microsoft.Compute/virtualMachines/acctest-VM-200929043203782538"
        ```

**Comment by mbfrahry** [comment]
Those tests have been failing on master as well. Thoughts on fixing it in another PR?

**Review by katbyte — APPROVED** [review]
+1

**Pull request merged**